TPC-H-based micro benchmarks

Bouncner opened the pull request:
I mentioned the topic a couple of weeks ago in the fortnightly.

I found myself too often writing the same micro benchmarks when working on the various joins. I thought it might make sense to create a skeleton microbenchmark which loads TPC-H tables and has a couple of useful benchmarks to have a quick starting point.
Generally, I think the micro benchmarks deserve a little bit more love.

This PR is supposed to get your opinion on such a micro benchmark. I like the benchmark framework, but I somewhat misuse it for the TPC-H micro benchmarks.

I added tests for the TPC-H Q6 scans, which might be handy for the SIMD PR (#1377) and for Jan's inclusive/exclusive-between group.
Further, there are benchmarks that show the need to improve our sort-merge join further since it might come in handy with rather complex join constructs (TPC-DS?!?).

Bouncner commented:
Oh, please ignore the commits concerning the hash join tests. That should not be part of this PR.

Bensk1 commented:
I agree that the microbenchmarks deserver a little bit more love. This came also up during the seminar because the groups have to misuse file-based or the TPC-H benchmark from time to time.

Bensk1 commented on src/benchmark/tpch_data_micro_benchmark.cpp:
```diff
@@ -0,0 +1,208 @@
+
+    if (!sm.has_table("lineitem")) {
+      std::cout << "Generating TPC-H data with scale factor " << scale_factor << " ... " << std::flush;
+      TpchDbGenerator(scale_factor, 100'000).generate_and_store();
```
The default parameter for the chunk size is `Chunk::DEFAULT_SIZE` which is `100'000`. You could not specify it at all or use `Chunk::DEFAULT_SIZE`.

Bensk1 commented on src/benchmark/tpch_data_micro_benchmark.cpp:
```diff
@@ -0,0 +1,208 @@
+      std::cout << "done." << std::endl;
+
+      // Tables are dictionary-encoded for the following benchmarks.
+      const auto default_encoding = EncodingType::Dictionary;
```
Should we print how the tables are encoded?

Bouncner commented:
Anything particular that would be helpful there?

For one of my groups, we plan to manually define LQP that are once with and once without nested queries to measure the impact of unnesting.

Bensk1 commented on src/benchmark/tpch_data_micro_benchmark.cpp:
```diff
@@ -0,0 +1,208 @@
+    _lineitem_wrapper = std::make_shared<TableWrapper>(lineitem_tab);
+    _lineitem_wrapper->execute();
+
+    _orders_wrapper = std::make_shared<TableWrapper>(orders_tab);
```
In case we start using more than two tables, this could be moved to a method that does this for all tables.

Bouncner commented:
Yes, let's discuss it tomorrow. Also, addressed your review.

Bouncner commented:
So is anybody against the idea of this PR? I think there have not been any objections during the discussion. Since I don't think we will have the "generate PQP-CPP code out of SQL queries" any time soon, I would continue with this PR and wait for your further reviews?

Bensk1 commented:
> Since I don't think we will have the "generate PQP-CPP code out of SQL queries" any time soon

@mrzzzrm, is it done yet?

Bouncner commented:
Ping, @mrzzzrm.

Done?

Bouncner commented on src/benchmark/tpch_data_micro_benchmark.cpp:
```diff
@@ -0,0 +1,217 @@
+        std::make_shared<BinaryPredicateExpression>(PredicateCondition::NotEquals, _lshipinstruct_operand, value_("a"));
+  }
+
+  void TearDown(::benchmark::State&) {}
```
Left there intentionally to avoid calling StorageManager.reset() each time a benchmark finished. Will add a note.

Bouncner commented:
WIP: two seminar groups asked for exemplary LQPs to work on. I think a good example of an unoptimized vs. optimized LQP would be beneficial for the students.

Bouncner commented:
Added two exemplary benchmarks which use LQPs. @mrzzzrm : does Mr. LQP accept these benchmarks?
I've addressed your other issues yesterday.

```➜  OpossumDB3 git:(martin/tpch_micro_benchmarks) ✗ ./rel/hyriseMicroBenchmarks --benchmark_filter="TPCHDataMicroBenchmarkFixture"
2019-01-18 15:25:26
Running ./rel/hyriseMicroBenchmarks
Run on (8 X 2900 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 262K (x4)
  L3 Unified 8388K (x1)
Generating TPC-H data set with scale factor 0.1 and Dictionary encoding:
- Loading/Generating tables (3 s 536 ms)
- Encoding tables (999 ms 180 µs)
- Adding Tables to StorageManager and generating statistics (870 ms 50 µs)
----------------------------------------------------------------------------------------------------------
Benchmark                                                                   Time           CPU Iterations
----------------------------------------------------------------------------------------------------------
TPCHDataMicroBenchmarkFixture/BM_TPCHQ6FirstScanPredicate             3135689 ns    3107217 ns        226
TPCHDataMicroBenchmarkFixture/BM_TPCHQ6SecondScanPredicate            1920948 ns    1918215 ns        362
TPCHDataMicroBenchmarkFixture/BM_TPCHQ6ThirdScanPredicate             1259855 ns    1248206 ns        579
TPCHDataMicroBenchmarkFixture/BM_TableScanIntegerOnPhysicalTable      4426947 ns    4416414 ns        157
TPCHDataMicroBenchmarkFixture/BM_TableScanIntegerOnReferenceTable     6608742 ns    6597433 ns         97
TPCHDataMicroBenchmarkFixture/BM_TableScanStringOnPhysicalTable       4424034 ns    4413270 ns        152
TPCHDataMicroBenchmarkFixture/BM_TableScanStringOnReferenceTable      6497946 ns    6485919 ns         99
TPCHDataMicroBenchmarkFixture/BM_TPCH4WithExistsSubquery           105116005884 ns 104680940000 ns          1
[PERF] Left relation larger than right relation hash join (input relations have been swapped). at src/lib/operators/join_hash.cpp:146
	Performance can be affected. This warning is only shown once.

TPCHDataMicroBenchmarkFixture/BM_TPCH4WithUnnestedSemiJoin           92811202 ns   92320667 ns          6
TPCHDataMicroBenchmarkFixture/BM_HashSemiProbeRelationSmaller       117145616 ns  116823200 ns          5
TPCHDataMicroBenchmarkFixture/BM_HashSemiProbeRelationLarger         33751201 ns   33620619 ns         21
TPCHDataMicroBenchmarkFixture/BM_SortMergeSemiProbeRelationSmaller   51085101 ns   50929182 ns         11
TPCHDataMicroBenchmarkFixture/BM_SortMergeSemiProbeRelationLarger    55211460 ns   55031917 ns         12
```

mrks commented on src/benchmark/CMakeLists.txt:
```diff
@@ -17,6 +17,7 @@ add_executable(
     operators/union_all_benchmark.cpp
     statistics/generate_table_statistics_benchmark.cpp
     tpch_table_generator_benchmark.cpp
+    tpch_data_micro_benchmark.cpp
```
sort plz
